src/memray/_memray.pyx

```diff
@@ -602,27 +614,104 @@ cdef class FileReader:
 
     @property
     def metadata(self):
-        def millis_to_dt(millis) -> datetime:
-            return datetime.fromtimestamp(millis // 1000).replace(
-                microsecond=millis % 1000 * 1000)
+        return _create_metadata(self._header, self._high_watermark.peak_memory)
+
+
+def compute_statistics(
+    file_name,
+    *,
+    report_progress=False,
+    num_largest=5,
+):
+    cdef shared_ptr[RecordReader] reader_sp = make_shared[RecordReader](
+        unique_ptr[FileSource](new FileSource(file_name))
+    )
+    cdef RecordReader* reader = reader_sp.get()
+
+    cdef header = reader.getHeader()
+    total = header["stats"]["n_allocations"] or None
+
+    cdef AllocationStatsAggregator aggregator
+    cdef ProgressIndicator progress_indicator = ProgressIndicator(
+        "Computing statistics",
+        total=total,
+        report_progress=report_progress,
+    )
+    with progress_indicator:
+        while True:
+            PyErr_CheckSignals()
+            ret = reader.nextRecord()
+            if ret == RecordResult.RecordResultAllocationRecord:
+                aggregator.addAllocation(reader.getLatestAllocation())
+                progress_indicator.update(1)
+            elif ret == RecordResult.RecordResultMemoryRecord:
+                pass
+            else:
+                break
 
-        stats = self._header["stats"]
-        allocator_id_to_name = {
-            PythonAllocatorType.PYTHON_ALLOCATOR_PYMALLOC: "pymalloc",
-            PythonAllocatorType.PYTHON_ALLOCATOR_PYMALLOC_DEBUG: "pymalloc debug",
-            PythonAllocatorType.PYTHON_ALLOCATOR_MALLOC: "malloc",
-            PythonAllocatorType.PYTHON_ALLOCATOR_OTHER: "unknown",
-        }
-        python_allocator = allocator_id_to_name[self._header["python_allocator"]]
-        return Metadata(start_time=millis_to_dt(stats["start_time"]),
-                        end_time=millis_to_dt(stats["end_time"]),
-                        total_allocations=stats["n_allocations"],
-                        total_frames=stats["n_frames"],
-                        peak_memory=self._high_watermark.peak_memory,
-                        command_line=self._header["command_line"],
-                        pid=self._header["pid"],
-                        python_allocator=python_allocator,
-                        has_native_traces=self._header["native_traces"])
+    # Ignore the n_allocations in the header, use our observed value.
+    header["stats"]["n_allocations"] = progress_indicator.num_processed
+
+    # NOTE: We need to cast away the constness of some references below,
+    #       because Cython's iterator handling is not const-correct.
+
+    # Convert allocation counts by allocator/by size to Python dicts.
+    allocation_count_by_allocator = {
```

Comment:
This is quite more expensive that it needs to be. This is creating an `AllocatorType` key and hashing it over for insertion. `AllocatorType` are enums in Python and those are quite expensive to build. It may be much faster to take the dictionary as it is (Cython automatically transforms C++ maps into dictionaries) and then create other where the keys are transformed.

Reply:
Discussed offline: this dict contains each allocator only once, so the cost of constructing this dict is absolutely trivial compared to all the work we just did reading the file.

The cost of hashing the `AllocatorType`, along with the cost of `+=` on a Python `int`, are the reason that I chose to build this up as a `unordered_map<int, size_t>` and only convert it to a Python `dict[AllocatorType, int]` at the very end.